dev/archery/archery/integration/tester_go.py

```diff
@@ -119,3 +131,124 @@ def flight_request(self, port, json_path=None, scenario_name=None):
         if self.debug:
             log(' '.join(cmd))
         run_cmd(cmd)
+
+    def make_c_data_exporter(self):
+        return GoCDataExporter(self.debug, self.args)
+
+    def make_c_data_importer(self):
+        return GoCDataImporter(self.debug, self.args)
+
+
+_go_c_data_entrypoints = """
+    const char* ArrowGo_ExportSchemaFromJson(const char* json_path,
+                                             uintptr_t out);
+    const char* ArrowGo_ImportSchemaAndCompareToJson(
+        const char* json_path, uintptr_t c_schema);
+
+    const char* ArrowGo_ExportBatchFromJson(const char* json_path,
+                                            int num_batch,
+                                            uintptr_t out);
+    const char* ArrowGo_ImportBatchAndCompareToJson(
+        const char* json_path, int num_batch, uintptr_t c_array);
+
+    int64_t ArrowGo_BytesAllocated();
+    void ArrowGo_RunGC();
+    void ArrowGo_FreeError(const char*);
+    """
+
+
+@functools.lru_cache
+def _load_ffi(ffi, lib_path=_INTEGRATION_DLL):
+    ffi.cdef(_go_c_data_entrypoints)
+    dll = ffi.dlopen(lib_path)
+    dll.ArrowGo_ExportSchemaFromJson
+    return dll
+
+
+class _CDataBase:
+
+    def __init__(self, debug, args):
+        self.debug = debug
+        self.args = args
+        self.ffi = cdata.ffi()
+        self.dll = _load_ffi(self.ffi)
+
+    def _pointer_to_int(self, c_ptr):
+        return self.ffi.cast('uintptr_t', c_ptr)
+
+    def _check_go_error(self, go_error):
+        """
+        Check a `const char*` error return from an integration entrypoint.
+
+        A null means success, a non-empty string is an error message.
+        The string is dynamically allocated on the Go side.
+        """
+        assert self.ffi.typeof(go_error) is self.ffi.typeof("const char*")
+        if go_error != self.ffi.NULL:
+            try:
+                error = self.ffi.string(go_error).decode('utf8',
+                                                         errors='replace')
+                raise RuntimeError(
+                    f"Go C Data Integration call failed: {error}")
+            finally:
+                self.dll.ArrowGo_FreeError(go_error)
+
+    def _run_gc(self):
+        self.dll.ArrowGo_RunGC()
+
+
+class GoCDataExporter(CDataExporter, _CDataBase):
+    # Note: the Arrow Go C Data export functions expect their output
+    # ArrowStream or ArrowArray argument to be zero-initialized.
+    # This is currently ensured through the use of `ffi.new`.
```

Comment:
huh, i thought `trampoline.c` zero-initialized everything but you're right, we're only trampolining to zero-initialize for `streamGetSchema` and `streamGetNext`, darn.

Reply:
I'm not sure what `trampoline.c` is, but I directly call into CGo-exported functions here.

Reply:
See https://github.com/apache/arrow-adbc/issues/729 for the full discussion, but essentially the callbacks that were populated into the C Data `ArrowArrayStream` struct for getting the schema and getting the next record batch had the same assumption that the output arguments were zero-initialized. So @lidavidm created `trampoline.c` which essentially provided wrappers around the Go exported `streamGetSchema` and `streamGetNext` function pointers, and we use the trampoline methods as the function pointers we use (see `exportStream` in `cdata/exports.go`)

But like i said, it looks like we only did that for the function pointers we set for the callbacks in the `ArrowArrayStream` struct and not for the other exported functions for regular import/export.

Reply:
It would be good to apply the same protections for the rest.

Notably, Arrow-C++ does NOT zero-initialize structs before using them so we should try to reflect reality here.

Reply:
I agree. I was just thinking it should be done in a different PR than this one, unless @pitrou feels like adding the same protections to the existing exported funcs here :smile:

Reply:
No, let's defer this to another issue and PR.